src/main/java/com/microsoftopentechnologies/windowsazurestorage/service/DownloadService.java

```diff
@@ -32,49 +31,72 @@
+    private Thread[] downloadThreads = new Thread[DOWNLOAD_THREAD_COUNT];
 
     public DownloadService(DownloadServiceData data) {
         super(data);
     }
 
-    protected int downloadBlobs(Iterable<ListBlobItem> blobItems)
-            throws URISyntaxException, StorageException, WAStorageException {
-        final DownloadServiceData serviceData = getServiceData();
-        int filesDownloaded = 0;
-        for (final ListBlobItem blobItem : blobItems) {
-            // If the item is a blob, not a virtual directory
-            if (blobItem instanceof CloudBlob) {
-                // Download the item and save it to a file with the same
-                final CloudBlob blob = (CloudBlob) blobItem;
-
-                // Check whether we should download it.
-                if (shouldDownload(
-                        serviceData.getIncludeFilesPattern(),
-                        serviceData.getExcludeFilesPattern(),
-                        blob.getName(),
-                        true)) {
-                    downloadBlob(blob);
-                    filesDownloaded++;
+    class DownloadThread implements Runnable {
+        @Override
+        public void run() {
+            while (!isScanFinished.get() || filesDownloaded.intValue() != filesNeedDownload) {
+                try {
+                    Object downloadItem = downloadItemDeque.poll(1, TimeUnit.SECONDS);
+                    if (downloadItem != null) {
+                        if (downloadItem instanceof CloudBlob) {
+                            downloadBlob((CloudBlob) downloadItem);
+                            filesDownloaded.addAndGet(1);
+                        } else if (downloadItem instanceof CloudFile) {
+                            downloadSingleFile((CloudFile) downloadItem);
+                            filesDownloaded.addAndGet(1);
+                        }
+                    }
+                } catch (InterruptedException | WAStorageException e) {
+                    final String message = Messages.AzureStorageBuilder_download_err(
+                            getServiceData().getStorageAccountInfo().getStorageAccName()) + ":" + e.getMessage();
+                    e.printStackTrace(error(message));
+                    println(message);
+                    setRunUnstable();
                 }
+            }
+        }
+    }
 
-            } else if (blobItem instanceof CloudBlobDirectory) {
-                final CloudBlobDirectory blobDirectory = (CloudBlobDirectory) blobItem;
-                if (shouldDownload(
-                        serviceData.getIncludeFilesPattern(),
-                        serviceData.getExcludeFilesPattern(),
-                        blobDirectory.getPrefix(),
-                        false)) {
-                    filesDownloaded += downloadBlobs(blobDirectory.listBlobs());
-                }
+    public void startDownloadThreads() {
+        for (int i = 0; i < DOWNLOAD_THREAD_COUNT; i++) {
```

Comment:
Does it mean we always try to start threads at the beginning no matter how many files to be download? It's wired that we start so many thread while we only want to download one or two files.

Reply:
Besides, Is it possible that we use ThreadPool to manage the threads?

Reply:
Make sense, but if we want to scanning and downloading at the same time, then we don't know how many files need to be downloaded at first. Should we finish the scanning process before we start downloading?


I will add ThreadPool to manage the threads.